pypesto/modelselection/unique_models.py

```diff
@@ -0,0 +1,111 @@
+#            [definition.split(PARAMETER_VALUE_DELIMITER)
+#                for definition in columns[PARAMETER_DEFINITIONS_START:]])
+#
+#def generate_model_definition_lines(
+#        model_name: str,
+#        sbml_filename: str,
+#        parameter_definitions: List[str]
+#) -> str:
+#    '''Yields all expanded model definitions.'''
+#    for index, selection in enumerate(
+#            itertools.product(*parameter_definitions)):
+#        yield model_name+f'_{index}' + '\t' + sbml_filename + '\t' + \
+#                '\t'.join(selection) + '\n'
+#
+#def unpack_file(file_name: str):
+#    '''
+#    Converts model definitions from the compressed form to the expanded form.
+#    '''
+#    expanded_models_file = tempfile.NamedTemporaryFile(mode='w', delete=False)
+#    with open(file_name) as fh:
+#        for line_index, line in enumerate(fh):
+#            if line_index != HEADER_ROW:
+#                for definition_line in generate_model_definition_lines(
+#                        *split_file_definition(line)):
+#                    expanded_models_file.write(definition_line)
+#            else:
+#                expanded_models_file.write(line)
+#    return expanded_models_file
+
+def _replace_estimate_symbol(parameter_definition) -> List:
+    return [ESTIMATE_SYMBOL_INTERNAL if p == ESTIMATE_SYMBOL_UI else p
+            for p in parameter_definition]
+
+def unpack_file(file_name: str):
+    '''
+    Unpacks a model definition file into a new temporary file that is returned.
+
+    TODO
+        - Consider alternatives to `_{n}` suffix for model `modelId`
+        - How should the selected model be reported to the user? Remove the
+          `_{n}` suffix and report the original `modelId` alongside the
+          selected parameters? Generate a set of PEtab files with the chosen
+          SBML file and the parameters specified in a parameter or condition
+          file?
+    '''
+    expanded_models_file = tempfile.NamedTemporaryFile(mode='w', delete=False)
+    with open(file_name) as fh:
+        for line_index, line in enumerate(fh):
+            if line_index != HEADER_ROW:
+                columns = line.strip().split('\t')
+                parameter_definitions = [
+                    _replace_estimate_symbol(
+                        definition.split(PARAMETER_VALUE_DELIMITER)
+                    ) for definition in columns[PARAMETER_DEFINITIONS_START:]
+                ]
+                for index, selection in enumerate(itertools.product(
+                        *parameter_definitions
+                )):
+                    expanded_models_file.write(
+                        '\t'.join([
+                            columns[MODEL_NAME_COLUMN]+f'_{index}',
+                            columns[SBML_FILENAME_COLUMN],
+                            *selection
+                        ]) + '\n'
+                    )
+            else:
+                expanded_models_file.write(line)
+    return expanded_models_file
```

Comment:
just for understanding: this now returns a file which has the model selection in a tab separated format and could be read by pandas data frame e.g. and loop over rows which our row2problem function could then convert to pypesto problems.
Correct?

Reply:
Yes the example:
| modelId | SBML      | parameterId1 | parameterId2 |
|---------|-----------|--------------|--------------|
| M1      | SBML1.xml | 0            | 0            |
| M2      | SBML1.xml | 1            | 0            |
| M3      | SBML2.xml | 0;1;-        | 0;1          |

is converted to:
| modelId | SBML      | parameterId1 | parameterId2 |
|---------|-----------|--------------|--------------|
| M1_0    | SBML1.xml | 0            | 0            |
| M2_0    | SBML1.xml | 1            | 0            |
| M3_0    | SBML2.xml | 0            | 0            |
| M3_1    | SBML2.xml | 0            | 1            |
| M3_2    | SBML2.xml | 1            | 0            |
| M3_3    | SBML2.xml | 1            | 1            |
| M3_4    | SBML2.xml | -            | 0            |
| M3_5    | SBML2.xml | -            | 1            |

and the temporary file object is returned. Do you think a data frame should be returned instead? If so, then the entire table would be stored in memory (at the moment, we intend to implement a generator, which would use less memory, but may be slower...).

Reply:
I think storing/returning in a (temp) file is fine. I would have just continued with reading this in as a dataframe since I find it easier to iterate over row while still keeping the head row.
But I guess there are mulitple ways to do this. Should we just discuss tomorrow morning?

Reply:
Sure, at the moment the generator will return each row as a dictionary, where the keys are the header elements, and the values are the row elements.